### PR TITLE
Add typing to manifest.package.name

### PR DESCRIPTION
**Pull request by fschutt**
This should fix problems where we constantly have to parse / unparse the `package.name` into a `namespace/name` format. Needed for https://github.com/wasmerio/wasmer/pull/3324.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -70,10 +70,52 @@ pub static README_PATHS: &[&str; 5] = &[
...
 pub static LICENSE_PATHS: &[&str; 3] = &["LICENSE", "LICENSE.md", "COPYING"];
 
+#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct PackageName {
+    pub namespace: String,
+    pub name: String,
+}
+
+impl PackageName {
+    pub fn parse(s: &str) -> Result<Self, &'static str> {
```
Comment (Michael-F-Bryan) [inline src/lib.rs:126]: Let's introduce a `ParsePackageNameError` type instead of using bare strings.

File: src/lib.rs [hunk 2]
```diff
@@ -70,10 +70,72 @@ pub static README_PATHS: &[&str; 5] = &[
...
+            .next()
+            .ok_or("no namespace in package name")?
+            .to_string();
+        Ok(Self { name, namespace })
+    }
+
+    /// Returns the `namespace/name` package name
+    pub fn to_string(&self) -> String {
+        format!("{}", self)
+    }
```
Comment (Michael-F-Bryan) [inline src/lib.rs:147]: This can be deleted because we get a `to_string()` method for free when implementing `std::fmt::Display`.

```suggestion
```

File: src/lib.rs [hunk 3]
```diff
@@ -70,10 +70,72 @@ pub static README_PATHS: &[&str; 5] = &[
...
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<PackageName, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let buf = String::deserialize(deserializer)?;
+        PackageName::parse(&buf).map_err(serde::de::Error::custom)
+    }
+}
```
Comment (Michael-F-Bryan) [inline src/lib.rs:182]: Can these be moved to `impl serde::Serialize for PackageName` and `impl<'de> serde::Deserialize<'de> for PackageName` blocks? Otherwise it'll be difficult for downstream crates to store a `PackageName` as a field.

File: src/lib.rs [hunk 4]
```diff
@@ -70,10 +70,72 @@ pub static README_PATHS: &[&str; 5] = &[
...
+    pub fn parse(s: &str) -> Result<Self, &'static str> {
+        if !s.contains('/') {
+            return Err("no / in package name");
+        }
+        let mut split = s.split('/');
+        let name = split.next().ok_or("no name in package name")?.to_string();
+        let namespace = split
+            .next()
+            .ok_or("no namespace in package name")?
+            .to_string();
```
Comment (Michael-F-Bryan) [inline src/lib.rs]: Ideally, we'd implement a proper parser (e.g. a mini recursive descent parser) for this instead of relying on `string.split()`. Otherwise we're going to miss lots of edge cases or mis-parse things.

File: src/lib.rs [hunk 5]
```diff
@@ -966,3 +1028,18 @@ annotations = { file = "Runefile.yml", kind = "yaml" }
         );
     }
 }
+
+#[test]
+fn test_package_name_parse() {
```
Comment (Michael-F-Bryan) [inline src/lib.rs:1083]: Let's add more test cases based on [the package naming rules doc](https://www.notion.so/wasmer/Naming-and-Ownership-Policy-ec6e449d8d5f4acd9a616ffa73612f5d).

Success:

- `namespace/package`
- `name-space/pack-age`
- `name_space/pack_age`
- `_/package`

Errors:
- `namespace/package/something-else` (only 2 segments)
- `name space/pack age` (i.e. no spaces)
- `namespace/`
- `/package`
- `namespace/_` (the name can't be `_`)
- `_namespace/_package` (can't have leading/trailing `_` or `-`)
- `namespace_/package_`
- `-namespace/-package`
- `namespace-/package-`
- `http://namespace`
- `nàméspàcë/pàckàgê` (i.e. only `-`, `_`, and ascii alphanumeric characters)
- (the empty string)
- `abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijz/abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijz` (i.e. namespace and packages are 101 characters when they can be at most 100 characters)

**Review by Michael-F-Bryan — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 4 above)
Comment (Michael-F-Bryan) [reply src/lib.rs]: This can be pushed off for a later refactor.

File: src/lib.rs (hunk 1 above)
Comment (fschutt) [reply src/lib.rs]: ok done

File: src/lib.rs (hunk 3 above)
Comment (fschutt) [reply src/lib.rs]: ok done

File: src/lib.rs (hunk 5 above)
Comment (fschutt) [reply src/lib.rs:1187]: ok done

**Review by Michael-F-Bryan — APPROVED** [review]
Awesome! Let's merge it :rocket:
